Build fixes for Windows and more

ocrete opened the pull request:
Here are some more build fixes part of my effort to build LCEVCdec as part of the GStreamer packages.

They mostly center on not confusing "Windows" and "MSVC" as GStreamer builds Windows targets also using mingw.

Also fix the install the pkg-config to use the install prefix as the pkg-config path is different on Debian and RedHat style distributions.

ocrete committed:
Use windows.h in small letters

The mingw toolchain is case sensitive when cross-compiling.

ocrete committed:
pkg-config: Use the CMake install prefix

Use it instead of assuming that its two levels below the pkgconfig,
Debian style distros install them in /usr/lib/$(arch)/pkgconfig

ocrete committed:
cmake: Leave the default install prefix as-is

ocrete committed:
Determine cpuid based on compiler instead of OS

ocrete committed:
Windows doesn't always mean MSVC as the compiler

For example, _lzcnt_u32() intrinsic is specific to MSVC, not Windows

ocrete committed:
Remove unused bitScanReverse

It's a pain to port, and its unused.

ocrete committed:
Make internal libs into OBJECT libraries

This way, they're bundled instead of installed when doing a static build

ocrete committed:
Also make the lcevc_dec_core library into an internal object library

ocrete committed:
Put licenses in documentation directory

ocrete commented:
I added 3 more things:
 * Build the internal libraires as "OBJECT" libraries instead of "STATIC", this way, when building the whole thing as a static library, it gets bundled into one library
 * Merge the lcevc_dec_core library into the lcevc_dec_api library, I'm not sure why they were split in the first place
 * Default to installing the licenses in the doc dir at /usr/share/doc/LCEVCdec/licenses

ocrete committed:
Put -lstdc++/-lc++ in the deps when building a static lib

And on Android, add -llog to access the Android logging functions

ocrete commented:
I pushed a few more comments:
 * When buliding a static library, put the deps into the .pc file. For example, on Android it's "-llog -lc++ -lm"
 * Install lcevc_config.h in the LCEVC subdir, to avoid filling /usr/include even more

ocrete commented:
With this, I have a working APK for Android out of GStreamer's CI

nanake commented:
The issue persists for me

```c
[root@2b786f5bed90 build]# ninja
[2/146] Building RC object src/api/CMakeFiles/lcevc_dec_api.dir/api_version.rc.res
FAILED: src/api/CMakeFiles/lcevc_dec_api.dir/api_version.rc.res
x86_64-w64-mingw32ucrt-windres -O coff  -I /lcevc/src/api/include -I /lcevc/src/api/src -I /lcevc/src/core/decoder/include -I /lcevc/src/core/sequencing/include -I /lcevc/src/api_utility/include -I /lcevc/src/api_utility/../api/include -I /lcevc/build/src/api -I /lcevc/build/generated -I /lcevc/include  /lcevc/build/src/api/api_version.rc src/api/CMakeFiles/lcevc_dec_api.dir/api_version.rc.res
x86_64-w64-mingw32ucrt-windres: no resources
[13/146] Building CXX object src/api/CMakeFiles/lcevc_dec_api.dir/src/log.cpp.obj
FAILED: src/api/CMakeFiles/lcevc_dec_api.dir/src/log.cpp.obj
/usr/sbin/x86_64-w64-mingw32ucrt-g++  -I/lcevc/src/api/include -I/lcevc/src/api/src -I/lcevc/src/core/decoder/include -I/lcevc/src/core/sequencing/include -I/lcevc/src/api_utility/include -I/lcevc/src/api_utility/../api/include -I/lcevc/build/src/api -isystem /lcevc/build/generated -isystem /lcevc/include -static-libgcc -static-libstdc++ -I/opt/ffbuild/include -O3 -pipe -D_FORTIFY_SOURCE=3 -fstack-protector-strong -fno-semantic-interposition -O3 -DNDEBUG -std=gnu++17 -mavx -Wall -Wshadow -Wwrite-strings --std=c++17 -MD -MT src/api/CMakeFiles/lcevc_dec_api.dir/src/log.cpp.obj -MF src/api/CMakeFiles/lcevc_dec_api.dir/src/log.cpp.obj.d -o src/api/CMakeFiles/lcevc_dec_api.dir/src/log.cpp.obj -c /lcevc/src/api/src/log.cpp
In file included from /lcevc/src/api/src/log.cpp:15:
/lcevc/src/api/src/log.h:164:73: error: 'uint32_t' has not been declared
  164 |     void print(LogComponent comp, LogLevel level, const char* function, uint32_t line,
      |                                                                         ^~~~~~~~
/lcevc/src/api/src/log.h:22:1: note: 'uint32_t' is defined in header '<cstdint>'; this is probably fixable by adding '#include <cstdint>'
   21 | #include <string>
  +++ |+#include <cstdint>
   22 |
/lcevc/src/api/src/log.h:166:74: error: 'uint32_t' has not been declared
  166 |     void printv(LogComponent comp, LogLevel level, const char* function, uint32_t line,
      |                                                                          ^~~~~~~~
/lcevc/src/api/src/log.h:166:74: note: 'uint32_t' is defined in header '<cstdint>'; this is probably fixable by adding '#include <cstdint>'
/lcevc/src/api/src/log.h:170:12: error: 'int64_t' does not name a type
  170 |     static int64_t getTicks(LogPrecision precision);
      |            ^~~~~~~
/lcevc/src/api/src/log.h:170:12: note: 'int64_t' is defined in header '<cstdint>'; this is probably fixable by adding '#include <cstdint>'
/lcevc/src/api/src/log.cpp:39:49: warning: 'thread' attribute directive ignored [-Wattributes]
   39 | static VNThreadLocal char sFormatBuffer[2][16384];
      |                                                 ^
/lcevc/src/api/src/log.cpp:143:6: error: no declaration matches 'void Logger::print(LogComponent, LogLevel, const char*, uint32_t, const char*, ...)'
  143 | void Logger::print(LogComponent comp, LogLevel level, const char* function, uint32_t line,
      |      ^~~~~~
/lcevc/src/api/src/log.h:164:10: note: candidate is: 'void Logger::print(LogComponent, LogLevel, const char*, int, const char*, ...)'
  164 |     void print(LogComponent comp, LogLevel level, const char* function, uint32_t line,
      |          ^~~~~
/lcevc/src/api/src/log.h:148:7: note: 'class Logger' defined here
  148 | class Logger
      |       ^~~~~~
/lcevc/src/api/src/log.cpp:152:6: error: no declaration matches 'void Logger::printv(LogComponent, LogLevel, const char*, uint32_t, const char*, va_list)'
  152 | void Logger::printv(LogComponent comp, LogLevel level, const char* function, uint32_t line,
      |      ^~~~~~
/lcevc/src/api/src/log.h:166:10: note: candidate is: 'void Logger::printv(LogComponent, LogLevel, const char*, int, const char*, va_list)'
  166 |     void printv(LogComponent comp, LogLevel level, const char* function, uint32_t line,
      |          ^~~~~~
/lcevc/src/api/src/log.h:148:7: note: 'class Logger' defined here
  148 | class Logger
      |       ^~~~~~
/lcevc/src/api/src/log.cpp:208:9: error: no declaration matches 'int64_t Logger::getTicks(LogPrecision)'
  208 | int64_t Logger::getTicks(LogPrecision precision)
      |         ^~~~~~
/lcevc/src/api/src/log.cpp:208:9: note: no functions named 'int64_t Logger::getTicks(LogPrecision)'
/lcevc/src/api/src/log.h:148:7: note: 'class Logger' defined here
  148 | class Logger
      |       ^~~~~~
/lcevc/src/api/src/log.cpp:100:13: warning: 'bool isSevere(LogLevel)' defined but not used [-Wunused-function]
  100 | static bool isSevere(LogLevel level)
      |             ^~~~~~~~
/lcevc/src/api/src/log.cpp:39:27: warning: 'sFormatBuffer' defined but not used [-Wunused-variable]
   39 | static VNThreadLocal char sFormatBuffer[2][16384];
      |                           ^~~~~~~~~~~~~
[15/146] Building CXX object src/api/CMakeFiles/lcevc_dec_api.dir/src/decoder.cpp.obj
ninja: build stopped: subcommand failed.
```



Adding `#include <cstdint>` in file `src/api/src/log.h` was a necessary step for [for GCC 15](https://gcc.gnu.org/gcc-15/porting_to.html), but the build is still failing.



```c
[root@2b786f5bed90 build]# ninja
[2/146] Building RC object src/api/CMakeFiles/lcevc_dec_api.dir/api_version.rc.res
FAILED: src/api/CMakeFiles/lcevc_dec_api.dir/api_version.rc.res
x86_64-w64-mingw32ucrt-windres -O coff  -I /lcevc/src/api/include -I /lcevc/src/api/src -I /lcevc/src/core/decoder/include -I /lcevc/src/core/sequencing/include -I /lcevc/src/api_utility/include -I /lcevc/src/api_utility/../api/include -I /lcevc/build/src/api -I /lcevc/build/generated -I /lcevc/include  /lcevc/build/src/api/api_version.rc src/api/CMakeFiles/lcevc_dec_api.dir/api_version.rc.res
x86_64-w64-mingw32ucrt-windres: no resources
[13/146] Building CXX object src/api/CMakeFiles/lcevc_dec_api.dir/src/log.cpp.obj
/lcevc/src/api/src/log.cpp:39:49: warning: 'thread' attribute directive ignored [-Wattributes]
   39 | static VNThreadLocal char sFormatBuffer[2][16384];
      |                                                 ^
/lcevc/src/api/src/log.cpp: In member function 'void Logger::printv(LogComponent, LogLevel, const char*, uint32_t, const char*, va_list)':
/lcevc/src/api/src/log.cpp:168:51: warning: format '%ld' expects argument of type 'long int', but argument 4 has type 'int64_t' {aka 'long long int'} [-Wformat=]
  168 |         count = snprintf(sFormatBuffer[1], 16384, "[%" PRId64 "]%s, %s (%u) - %s",
      |                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  169 |                          getTicks(m_timingPrecision), compName, function, line, sFormatBuffer[0]);
      |                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                  |
      |                                  int64_t {aka long long int}
In file included from /lcevc/src/api/src/log.cpp:15:
/lcevc/src/api/src/log.h:42:18: note: format string is defined here
   42 | #define PRId64 "ld"
[15/146] Building CXX object src/api/CMakeFiles/lcevc_dec_api.dir/src/decoder.cpp.obj
ninja: build stopped: subcommand failed.
```

MartinEesmaa commented:
Hi, @nanake!

Thanks for bringing up your little fix about cstdint! Here is another fix to change from WIN32 to MSVC:
https://github.com/v-novaltd/LCEVCdec/blob/aae588337b9de3a181f363101f319682b8852bd8/cmake/modules/VNovaSetup.cmake#L127-L132

This only affects the problem for GCC when building a Windows executable.
I'm gonna fix my forked repo.

Please let me know if you have any questions or feedback. Thanks! :)

- Martin Eesmaa

nanake commented:
Thanks, that fixes the problem for me.

Anyway, it cannot build with ffmpeg. I got `undefined reference to `__imp_LCEVC_CreateDecoder'` 🤔

```c
require_pkg_config liblcevc_dec lcevc_dec >= 2.0.0 LCEVC/lcevc_dec.h LCEVC_CreateDecoder
check_pkg_config liblcevc_dec lcevc_dec >= 2.0.0 LCEVC/lcevc_dec.h LCEVC_CreateDecoder
test_pkg_config liblcevc_dec lcevc_dec >= 2.0.0 LCEVC/lcevc_dec.h LCEVC_CreateDecoder
pkg-config --exists --print-errors lcevc_dec >= 2.0.0
check_func_headers LCEVC/lcevc_dec.h LCEVC_CreateDecoder -I/opt/ffbuild/include -L/opt/ffbuild/lib -lstdc++ -lm -llcevc_dec_api
test_ld cc -I/opt/ffbuild/include -L/opt/ffbuild/lib -lstdc++ -lm -llcevc_dec_api
test_cc -I/opt/ffbuild/include -L/opt/ffbuild/lib
BEGIN /tmp/ffconf.8av5oBGO/test.c
    1	#include <LCEVC/lcevc_dec.h>
    2	#include <stdint.h>
    3	long check_LCEVC_CreateDecoder(void) { return (long) LCEVC_CreateDecoder; }
    4	int main(void) { int ret = 0;
    5	 ret |= ((intptr_t)check_LCEVC_CreateDecoder) & 0xFFFF;
    6	return ret; }
END /tmp/ffconf.8av5oBGO/test.c
x86_64-w64-mingw32ucrt-gcc -D_ISOC11_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -U__STRICT_ANSI__ -D__USE_MINGW_ANSI_STDIO=1 -D__printf__=__gnu_printf__ -D_POSIX_C_SOURCE=200112 -D_XOPEN_SOURCE=600 -DPIC -static-libgcc -static-libstdc++ -I/opt/ffbuild/include -O3 -pipe -D_FORTIFY_SOURCE=3 -fstack-protector-strong -fno-semantic-interposition -DLIBTWOLAME_STATIC -std=c17 -fomit-frame-pointer -I/opt/ffbuild/include -pthread -I/opt/ffbuild/include -DWITH_GZFILEOP -I/opt/ffbuild/include -DCHROMAPRINT_NODLL -I/opt/ffbuild/include/oapv -DOAPV_STATIC_DEFINE -DOAPV_STATIC_DEFINE -I/opt/ffbuild/include -I/opt/ffbuild/include/freetype2 -DWITH_GZFILEOP -I/opt/ffbuild/include -I/opt/ffbuild/include/fribidi -DFRIBIDI_LIB_STATIC -I/opt/ffbuild/include/harfbuzz -I/opt/ffbuild/include/freetype2 -DWITH_GZFILEOP -I/opt/ffbuild/include/libxml2 -DLIBXML_STATIC -I/opt/ffbuild/include -I/opt/ffbuild/include -I/opt/ffbuild/include -I/opt/ffbuild/include -I/opt/ffbuild/include/dvdcss -I/opt/ffbuild/include -I/opt/ffbuild/include/dvdcss -I/opt/ffbuild/include -I/opt/ffbuild/include -I/opt/ffbuild/include/freetype2 -DWITH_GZFILEOP -I/opt/ffbuild/include/libxml2 -DLIBXML_STATIC -I/opt/ffbuild/include/freetype2 -I/opt/ffbuild/include -DWITH_GZFILEOP -I/opt/ffbuild/include/fribidi -DFRIBIDI_LIB_STATIC -I/opt/ffbuild/include/harfbuzz -I/opt/ffbuild/include/freetype2 -I/opt/ffbuild/include -DWITH_GZFILEOP -I/opt/ffbuild/include -I/opt/ffbuild/include -DHWY_STATIC_DEFINE -pthread -DJXL_STATIC_DEFINE -DJXL_CMS_STATIC_DEFINE -I/opt/ffbuild/include -DJXL_THREADS_STATIC_DEFINE -I/opt/ffbuild/include -DKVZ_STATIC_LIB -I/opt/ffbuild/include -I/opt/ffbuild/include -L/opt/ffbuild/lib -c -o /tmp/ffconf.8av5oBGO/test.o /tmp/ffconf.8av5oBGO/test.c
/tmp/ffconf.8av5oBGO/test.c: In function 'check_LCEVC_CreateDecoder':
/tmp/ffconf.8av5oBGO/test.c:3:47: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
    3 | long check_LCEVC_CreateDecoder(void) { return (long) LCEVC_CreateDecoder; }
      |                                               ^
x86_64-w64-mingw32ucrt-gcc -static-libgcc -static-libstdc++ -L/opt/ffbuild/lib -O3 -pipe -fstack-protector-strong -Wl,--nxcompat,--dynamicbase -Wl,--high-entropy-va -Wl,--as-needed -Wl,--pic-executable,-e,mainCRTStartup -Wl,--image-base,0x140000000 -I/opt/ffbuild/include -L/opt/ffbuild/lib -o /tmp/ffconf.8av5oBGO/test.exe /tmp/ffconf.8av5oBGO/test.o -lstdc++ -lm -llcevc_dec_api
/usr/lib/gcc/x86_64-w64-mingw32ucrt/15.0.1/../../../../x86_64-w64-mingw32ucrt/bin/ld: /tmp/ffconf.8av5oBGO/test.o:test.c:(.text+0x2): undefined reference to `__imp_LCEVC_CreateDecoder'
collect2: error: ld returned 1 exit status
ERROR: lcevc_dec >= 2.0.0 not found using pkg-config
```

MartinEesmaa commented:
Same for me too... :(
https://github.com/MartinEesmaa/FFmpeg-Builds/actions/runs/15248723358/job/42888665106

We have to wait a little bit until someone fixes the problem.

- Martin Eesmaa

MartinEesmaa commented:
@nanake, I have great news.

I found that linking dllimport and dllexport could only cause a static library problem, so I removed them and added a flag called `-DVNEnablePublicAPIExport` to avoid undefined references.

I fixed my problem🙂, so I released new commits to my forked repo, which you can look at.

https://github.com/v-novaltd/LCEVCdec/blob/aae588337b9de3a181f363101f319682b8852bd8/include/LCEVC/lcevc.h#L30-L33

I removed the dllspec of dllimport & dllexport to avoid problems, so that is my critical thinking analysis.
Also, it needs to be fully fixed by swapping libs, not this:

Original produced:
```
Libs: -L"${libdir}" -lstdc++ -lm -llcevc_dec_api 
```

To fix manually:

```
Libs: -L"${libdir}" -llcevc_dec_api -lstdc++ -lm
```

So, my MPEG-5 LCEVC decoder script has successfully compiled on all builds except winarm64.
Script: https://github.com/MartinEesmaa/FFmpeg-Builds/blob/master/scripts.d/50-lcevcdec.sh

This script fixes the pkg-config after the Ninja install by creating a new pkg-config file and adding flags for defining the Public API export.

Actions job: https://github.com/MartinEesmaa/FFmpeg-Builds/actions/runs/15263022639/job/42925392837

- Martin Eesmaa

walisiewicz-V-Nova commented:
Hi all, thanks for your contributions to this issue. We appreciate the requirement for Windows builds via mingw-gcc and this will be supported at least for a '[libraries only](https://github.com/v-novaltd/LCEVCdec/blob/main/docs/building.md#building-libraries-only---native)' build by V-Nova from the next release onwards.

@ocrete would you be able to remove the partial changes related to mingw-gcc build and we can tackle the [remaining changes](https://github.com/v-novaltd/LCEVCdec/pull/23#issuecomment-2855530623) in this PR more easily?

ocrete committed:
lint: Don't assume that the source branch is on the origin

It would be expected to be in the contributor's repo, or maybe not
even pushed

ocrete committed:
lint: Access different micro-rev of autopep8

Fedora ships 2.3.2 instead of 2.3.1, it should be fine I think

ocrete committed:
Add cstdint to log.h

This seems to be required with some GCC versions.

ocrete commented:
I updated my branch with the following changes:
* Fixed the lint.py script to not assume that branches are in origin/
* Accept autopep8 2.3.x and not just 2.3.1
* Ran the lint.py on every commit
* Move the PC_EXTRA_LIBS setting to VNovaProject.cmake instead of putting in the main CMakeLists.txt file, putting it in VNovaSetup.cmake doesn't work as it requires variables that haven't been set already at that point
* Removed the location change of lcevc_config.h, as it seems this is already done in an internal branch
* Added the <cstdint>

I see that Martin has just skipped the .rc file generation entirely, but I'm guessing this isn't actually correct and they should be generated based on static information when not building from a git release

walisiewicz-V-Nova commented:
@ocrete Closing this PR as these updates are included in [release 3.3.8](https://github.com/v-novaltd/LCEVCdec/releases/tag/3.3.8) with the exception of the change of building static libraries as object files which has been resolved by providing a pkg-config file that includes all the required static libraries in the correct order.
